design/05_data_structures_proposal.md

```diff
@@ -0,0 +1,141 @@
+# Data structures proposal
+
+A few low behavior data structures and traits need to be defined to be used as
+input and output of interfaces. This proposal attempt to specify them.
+
+## Backup requirement class
+
+The user can specify for each project what is his target for backups. For low
+importance projects, he might want a single copy on a secondary partition on the
+same physical device, while for highly sensitive data he might want to set a
+higher target. We may specify such target with the structure:
+
+```rs
+enum SecurityLevel {
+    // Connected to network, no authorization required
+    NetworkPublic,              // referenced, accessible to anyone
+    NetworkUnreferenced,        // unreferenced, accessible to anyone
+
+    // Connected to network, authorization required
+    NetworkUntrustedRestricted, // untrusted provider (e.g. under the PATRIOT Act)
+    NetworkTrustedRestricted,   // trusted provider
+
+    // Connected to a local only network
+    NetworkLocal,               // local network, no internet access
+
+    // Disconnected from network
+    Local,                      // local location
+    LocalMaxSecurity,           // local location with maximum security (in a safe?)
+}
+
+struct BackupRequirementClass {
+    // The number of distinct copies we want to keep (including the primary)
+    target_copies: u32 = 3,
+
+    // The number of distinct physical locations we want to keep the copies
+    target_locations: u32 = 1,
+
+    // The minimum security level of the backups
+    min_security_level: SecurityLevel = SecurityLevel::NetworkUntrustedRestricted,
+
+    // Name of the backup requirement class
+    name: String,
+}
+```
+
+## Project status
+
+A project might have several statuses:
+
+```rs
+enum ProjectTrackingStatus {
+    // The project is explicitly not tracked, as specified in the global config
+    Ignored,
+
+    // The project is explicitly tracked, it has a config file
+    Tracked,
+
+    // The project is not yet specified. This is a risk and the user should
+    // eventually specify it.
+    Unspecified,
+}
+
+struct ProjectBackupStatus {
```

Comment:
is that stored on the primary device or the secondary
if it is on the primary, shouldn't it be linked to the project in some way ?

Reply:
As the secondary might not be connected, it shall be on the primary. It might be on a global history file similar to a .bash_history, referencing a unique project id. This way we limit what is in the project config file only to config.